anonymizer/OpenDiffix.Service/Led.fs

```diff
@@ -0,0 +1,176 @@
+  let aggFns, aggArgs = aggregators |> Array.ofList |> Executor.Utils.unpackAggregators
+  let groupingLabelsLength = groupingLabels.Length
+  let rowSize = groupingLabelsLength + aggFns.Length
+
+  let lowCountIndex = findAggregator DiffixLowCount aggFns
+  let countIndex = findAggregator DiffixCount aggFns
+
+  let mergeBucketInto = mergeBucket groupingLabelsLength lowCountIndex countIndex
+
+  let makeBucket (group: Value array) =
+    let bucketSeed = Executor.Utils.addValuesToSeed executionContext.NoiseLayers.BucketSeed group
+
+    {
+      Aggregators = aggFns |> Array.map (Aggregator.create executionContext LOCAL_AGGREGATION)
+      LowCountArguments = Collections.Generic.List<Value list>()
+      CountArguments = Collections.Generic.List<Value list>()
+      ExecutionContext =
+        { executionContext with
+            NoiseLayers = { executionContext.NoiseLayers with BucketSeed = bucketSeed }
+        }
+      // Aggregator values are computed later
+      OutputRow = Array.init rowSize (fun i -> if i < groupingLabelsLength then group.[i] else Null)
+      IsLowCount = false
+    }
+
+  let state = Collections.Generic.Dictionary<Row, BucketState>(Row.equalityComparer)
+
+  // Group rows into buckets and keep track of low-count/count arguments
+  for row in Executor.execute executionContext childPlan do
+    let group = groupingLabels |> Array.map (Expression.evaluate row)
+    let argEvaluator = Expression.evaluate row
+
+    let bucket =
+      match state.TryGetValue(group) with
+      | true, bucketState -> bucketState
+      | false, _ ->
+        let bucketState = makeBucket group
+        state.[group] <- bucketState
+        bucketState
+
+    bucket.Aggregators
+    |> Array.iteri (fun i aggregator ->
+      let args = aggArgs.[i] |> List.map argEvaluator
+
+      if i = lowCountIndex then bucket.LowCountArguments.Add(args)
+      else if i = countIndex then bucket.CountArguments.Add(args)
+
+      args |> aggregator.Transition
+    )
+
+  // Compute aggregator final values
+  let buckets =
+    state
+    |> Seq.map (fun pair ->
+      let bucket = pair.Value
+      let executionContext = bucket.ExecutionContext
+      let outputRow = bucket.OutputRow
+
+      bucket.Aggregators
+      |> Array.iteri (fun i aggregator ->
+        if i = lowCountIndex then
+          let isLowCountValue = aggregator.Final executionContext
+          bucket.IsLowCount <- Value.unwrapBoolean isLowCountValue
+          outputRow.[groupingLabelsLength + i] <- isLowCountValue
+        else
+          outputRow.[groupingLabelsLength + i] <- aggregator.Final executionContext
+      )
+
+      bucket
+    )
+    |> Seq.toArray
+
+  // Merge victim buckets into sibling buckets
+  buckets
+  |> Array.iteri (fun i victimBucket ->
+    if victimBucket.IsLowCount then
+      let victimRow = victimBucket.OutputRow
+      let siblingsPerColumn = Array.init groupingLabelsLength (fun _i -> Collections.Generic.List<BucketState>(2))
```

Comment:
Why is it worth using `Collections.Generic.List` here instead of the native `list` or `array`?

Reply:
We need to keep track of how many buckets are matching. This could as well be a discriminated union with 3 cases: `NoBucket | SingleBucket bkt | MultipleBuckets`

Reply:
Either that or a regular list/array. I find the usage of `Collections.Generic.List` confusing.

Reply:
I will convert to union because it should be much more readable.